### PR TITLE
Feature spec testing

### PR DESCRIPTION
**Pull request by bmallred**
# Epics

 - #302: Feature Spec Testing

# Issues

 - #322: Create docker image for feature specs to execute
 - #323: Setup the directory structure needed for cucumber
 - #324: Create an example feature test logging on to the system

## General Architecture

Allow feature specs should be stored within `specs/` following the typical `cucumber` directory structure. For more information please refer to issue #322 and #323.

## Artifacts

All artifacts are kept within the `specs/` directory but ignored in `git`

 - Reports will be written to `specs/reports`
 - Screenshots will be written to `specs/screenshots`

## Changes to pre-existing technology and configurations

 - It was necessary to allow the 2FA token to be reset due to the nature of the testing. This is authorized using the `ALLOW_2FA_RESET` environment variable. If a value is present then it is enabled.
 - The `caddy` (used for the `web` service in docker) configuration file was altered to proxy API requests through the same HTTP server transparently to the backend server itself. By doing this we minimize what is exposed to the host and allow one entry point in to the web application.
 - `docker-compose` was updated to leverage version 2 taking advantage of networking features. This was done so a consistent experience was maintained regardless to where the browser resides.
 - In most development/testing environments the environment variable `API_BASE_URL` is no longer necessary. The use of this variable should only be when the API backend is separated from the frontend by URIs.

### CODE REVIEW

File: specs/features/step_definitions/authentication.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    return client
+      .assert.urlContains('/login')
+      .assert.visible('#twofactor-component')
+      .isVisible('#twofactor-component', (result) => {
+        // If the QR code is not present then we need to reset two factor
+        // authentication to continue with our flow.
+        client.click('.reset').pause(2000)
+      })
+      .saveScreenshot(screenshot, () => {
```
Comment (adelevie) [inline specs/features/step_definitions/authentication.js]: @bmallred looks great! Any way to have screenshots saved for all features? And maybe even for each step?

Comment (bmallred) [inline specs/features/step_definitions/authentication.js]: Yeah absolutely.

So this example feature was just for logging in to the system. In future features I would assume they all have to do this so would this be compressed in to one step then?

File: specs/features/step_definitions/authentication.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      .waitForElementVisible('body', 1000)
+  })
+
+  Then(/^sign in$/, () => {
+    return client
+      .setValue('input[type="text"]', credentials.username)
+      .setValue('input[type="password"]', credentials.password)
+      .click('#basic button[type="submit"]')
+      .pause(3000)
+      // .getLog('browser', (entries) => {
```
Comment (billy22g) [inline specs/features/step_definitions/authentication.js]: These lines of commented out code don't seem to be needed anymore. Is that accurate?

Comment (bmallred) [inline specs/features/step_definitions/authentication.js]: Correct. Left them in the example in the case someone wanted a reference for debugging. We can take them out though b/c there are examples online.

**Comment by bmallred** [comment]
authentication

**Pull request merged**